File: Practica1_i-Lunch/src/gestorAplicacion/usuariosRestaurante/Administrador.java
package gestorAplicacion.usuariosRestaurante;

import java.io.Serializable;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.Random;

import gestorAplicacion.gestionRestaurante.*;

/* Esta es la clase principal de la aplicación, ya que el administrador
 * es quien tiene acceso a la aplicación y desde su cuenta se manejan
 * todas las funcionalidades. Esta clase hereda de Empleado */

public class Administrador extends Empleado implements Serializable, Usuario {
	/*
	 * Como Administrador hereda de Empleado, se utilizarán los atributos de este.
	 * Solo se implementa el atributo extra IMPUESTOS ya que se usará en una de sus
	 * funcionalidades
	 */
	private static final float IMPUESTOS = 1.19f; // 19% de impuestos

	// Atributos utilizados para la serialización
	private static final long serialVersionUID = 1L; // Se requiere del atributo serialVersionUID por usar la interface
														// Serializable.
	private static ArrayList<Administrador> administradores;
	static {
		administradores = new ArrayList<Administrador>();
	}

	// Constructor de la clase Administrador. No se recibe el atributo "cargo"
	// debido a que este siempre será "Administrador"
	public Administrador(int cedula, String nombre, boolean disponibilidad, int salario, Restaurante restaurante) {
		super(cedula, nombre, "Administrador", disponibilidad, salario, restaurante);
		administradores.add(this);
	}

	// Sobrecarga del constructor. No toma parámetros y llama al constructor usando
	// valores predeterminados.
	public Administrador() {
		this(0, "NN", false, 0, null);
	}

	// Método utilizado para la serialización/deserielización
	public static ArrayList<Administrador> getAdministradores() {
		return administradores;
	}

	// Getter de IMPUESTOS (No se implementa Setter debido a que es una constante
	public static float getImpuestos() {
		return Administrador.IMPUESTOS;
	}

	/*
	 * Este método recibe como parámetro un objeto Empleado, comprueba que no
	 * esté ya contratado en el restaurante y si no lo está lo agrega a la lista
	 * de empleados del restaurante. También cambia el atributo Resturante del
	 * Empleado.
	 */
	public void contratarEmpleado(Empleado empleado) {
		ArrayList<Empleado> listaEmpleados = this.restaurante.getEmpleados();
		listaEmpleados.add(empleado);
		this.restaurante.setEmpleados(listaEmpleados);
	}

	/*
	 * Este método recibe como parámetro un objeto Empleado, comprueba que este
	 * esté ya contratado en el restaurante y si lo está lo agrega a la lista de
	 * empleados del restaurante. También cambia el atributo Resturante del
	 * Empleado.
	 */
	public void despedirEmpleado(Empleado empleado) {
		ArrayList<Empleado> listaEmpleados = this.restaurante.getEmpleados();
		listaEmpleados.remove(empleado);
		this.restaurante.setEmpleados(listaEmpleados);
	}

	/*
	 * Este método recibe como parámetros: - Un String con el nombre del producto
	 * - Un String con la descripción del producto - Un int con el precio del
	 * procucto - Un bool que es true si el producto solo puede ser vendido a
	 * mayores de edad - Un bool que indica la disponibilidad actual del producto
	 * Luego, se crea un objeto de tipo Producto con estos atributos y se agrega a
	 * la lista de productos del restaurante. Antes de agregarlo se comprueba que no
	 * haya un producto con un nombre exactamente igual y si esto no se cumple, se
	 * agrega.
	 */
	public String crearProducto(String nombre, String descripcion, int precio, boolean disponibilidad,
			boolean restriccion, int cantidad) {
		Producto productoNuevo = new Producto(nombre, descripcion, precio, disponibilidad, restriccion, cantidad);

		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		if (!listaMenu.contains(productoNuevo)) {
			listaMenu.add(productoNuevo);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + nombre + " creado con éxito";
		} else {
			return "ERROR: El producto ya se encuentra creado";
		}
	}

	/*
	 * Método útil para cambiar el nombre de un Producto pasado por parámetro.
	 * Antes de cambiar el atributo se debe de comprobar que el producto esté en la
	 * lista de productos del restaurante.
	 */
	public String actualizarNombreProducto(int producto, String nombre) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setNombre(nombre);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch (Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método útil para cambiar la descripción de un Producto pasado por
	 * parámetro. Antes de cambiar el atributo se debe de comprobar que el producto
	 * esté en la lista de productos del restaurante.
	 */
	public String actualizarDescripcionProducto(int producto, String descripcion) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setDescripcion(descripcion);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch (Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método útil para cambiar el precio de un Producto pasado por parámetro.
	 * Antes de cambiar el atributo se debe de comprobar que el producto esté en la
	 * lista de productos del restaurante.
	 */
	public String actualizarPrecioProducto(int producto, int precio) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setPrecio(precio);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch(Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método útil para cambiar el estado de la restricción de un Producto pasado
	 * por parámetro. Antes de cambiar el atributo se debe de comprobar que el
	 * producto esté en la lista de productos del restaurante.
	 */
	public String actualizarRestriccionProducto(int producto, boolean restriccion) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setRestriccion(restriccion);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch(Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método útil para cambiar la disponibilidad actual de un Producto pasado por
	 * parámetro. Antes de cambiar el atributo se debe de comprobar que el producto
	 * esté en la lista de productos del restaurante.
	 */
	public String actualizarDisponibilidadProducto(int producto, boolean disponibilidad) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setDisponiblidad(disponibilidad);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch(Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método útil para cambiar la cantidad actual de un Producto pasado por
	 * parámetro. Antes de cambiar el atributo se debe de comprobar que el producto
	 * esté en la lista de productos del restaurante.
	 */
	public String actualizarCantidadProducto(int producto, int cantidad) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		try {
			Producto productoActualizado = listaMenu.get(producto);
			productoActualizado.setCantidad(cantidad);
			listaMenu.set(producto, productoActualizado);
			this.restaurante.setMenu(listaMenu);
			return "Producto " + productoActualizado.getNombre() + " actualizado con éxito";
		} catch(Exception e) {
			return "ERROR: El producto que intentas actualizar no existe";
		}
	}

	/*
	 * Método que sirve para eliminar un Producto de la lista de productos del
	 * restaurante. Antes de eliminar el producto se debe de verificar que el
	 * producto se encuentre en la lista de productos del restaurante.
	 */
	public String eliminarProducto(String producto) {
		ArrayList<Producto> listaMenu = this.restaurante.getMenu();
		int productoID;
		try {
			productoID = Integer.parseInt(producto);
			listaMenu.remove(productoID);
			this.restaurante.setMenu(listaMenu);
			return "Producto con ID: " + productoID + " eliminado con éxito";
		} catch(Exception e) {
			return "ERROR: El producto que intentas eliminar no existe";
		}
	}

	/*
	 * Método que realiza el pago de la nómina a todos los empleados del
	 * restaurante incluyendo al Administrador mismo. Antes de hacer efectivo el
	 * pago se debe de comprobar que en el balance de cuenta del restaurtante
	 * existan fondos sufucientes para pagar a los empleados
	 */
	public String pagoNomina() {
		ArrayList<Empleado> listaEmpleados = this.restaurante.getEmpleados();
		float totalSalarios = 0;

		for (Empleado empleado : listaEmpleados) {
			totalSalarios += (empleado.getSalario() * Administrador.getImpuestos()); // Salario más un impuesto
		}

		if (totalSalarios <= this.restaurante.getBalanceCuenta()) {
			float nuevoBalance = this.restaurante.getBalanceCuenta() - totalSalarios;
			this.restaurante.setBalanceCuenta(nuevoBalance);
			return "Nómina de todos los empleados pagada con éxito. El nuevo balance de cuenta es: $"
					+ restaurante.getBalanceCuenta();
		} else {
			return "ERROR: No se posee el suficiente dinero para pagar la nómina de todos los empleados";
		}
	}

	/*
	 * Método que realiza el pago de la nómina a un solo los empleado del
	 * restaurante. Antes de hacer efectivo el pago se debe de comprobar que en el
	 * balance de cuenta del restaurtante existan fondos sufucientes para pagar al
	 * empleado en cuestión
	 */
	public String pagoNomina(Empleado empleado) {
		if ((empleado.getSalario() * Administrador.getImpuestos()) <= this.restaurante.getBalanceCuenta()) { // Salario
																												// más
																												// un
																												// impuesto
			float nuevoBalance = this.restaurante.getBalanceCuenta()
					- (empleado.getSalario() * Administrador.getImpuestos());
			this.restaurante.setBalanceCuenta(nuevoBalance);
			return "Nómina de todos los empleados pagada con éxito. El nuevo balance de cuenta es: $"
					+ restaurante.getBalanceCuenta();
		} else {
			return "ERROR: No se posee el suficiente dinero para pagar la nómina de todos los empleados";
		}
	}

	/*
	 * Método que sirve para simular un pedido al restaurante. Este recibe como
	 * parámetros un Cliente con quien se asociará el pedido simulado.
	 */
	public Pedido simularPedido(Cliente cliente) {
		Pedido pedido = new Pedido();
		int idTipo = (int) (Math.random()*2);
		String tipo = "";

		// Se asigna el cliente
		pedido.setCliente(cliente);
		// Se asigna el restaurante
		pedido.setRestaurante(restaurante);
		// Se le asigna un codigo
		pedido.setCodigo(Pedido.getTotalPedidos());
		// Se guarda fecha y hora
		pedido.setFechaHora(LocalDateTime.now());
		// Se le asigna un tipo 
		switch (idTipo) {
		case 0: {
			tipo = TipoPedido.Llevar.toString();
		}
		case 1: {
			tipo = TipoPedido.EnTienda.toString();
		}
		case 2: {
			tipo = TipoPedido.Domicilio.toString();
		}
		}
		pedido.setTipo(tipo);
		
		
	// Se asgina la lista de productos
		// Se define la canidad de productos a ordenar con base en la cantidad de productos disponible en el restaurante.
		// De 1 a la cantidad disponible en el restaurante.
		int cantidadProductos = (int) ((Math.random()*(restaurante.getMenu().size()-1))+1);
		ArrayList<Producto> productos = new ArrayList<Producto>();
		//En un ciclo que se repite el numero de productos a solicitar se seleccionan los productos del menu del restaurante.
		for(int i=0; i<cantidadProductos;i++){
			// Generamos un numero aleatorio para seleccionar un producto al azar de entre el menu del restaurante.
			int idProducto = (int) (Math.random() * restaurante.getMenu().size());
			Producto producto = restaurante.getMenu().get(idProducto);
			// Si el producto ya hace parte del pedido no se agrega
			if(!productos.contains(producto)) {
				// Se genera una cantidad aleatoria del producto seleccionado para ser agregado al pedido.
				int cantidadXProducto = (int) ((Math.random()*(5-1))+1);
				producto.setCantidad(cantidadXProducto);
				productos.add(producto);
			}else {
				//Si no se selecciona un producto se repite el ciclo.
				i--;
			}
		}
		pedido.setProductos(productos);
		
		 //Se calcula el precio total 
		int total = 0;
		for(Producto producto:pedido.getProductos()) {
			total += producto.getCantidad() * producto.getPrecio();
		}
		pedido.setPrecioTotal(total);
		//Se pone en estado de Enviado
		pedido.setEstado(estadoPedido.Enviado.toString());
		//Se agregan los mensajes del usuario
		//Se retorna el pedido completo
		restaurante.agregarPedido(pedido);
		return pedido;

	}

	// Implementación de la interfaz Usuario
	public String informacion() {
		if (this.getDisponibilidad()) {
			return "El Administrador del restaurante + " + this.restaurante.getNombre() + " es " + this.nombre
					+ " con C.C. " + this.cedula + ".\n" + "Tiene un salario de: $" + this.salario + "\n"
					+ "Está disponible actualmente.";
		} else {
			return "El Administrador del restaurante + " + this.restaurante.getNombre() + " es " + this.nombre
					+ " con C.C. " + this.cedula + ".\n" + "Tiene un salario de: $" + this.salario + "\n"
					+ "No está disponible actualmente.";
		}
	}
}

File: Practica2_i-Lunch/src/gestorAplicacion/usuariosRestaurante/administrador.py

# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS
# ! ACTUALIZAR METODOS DE ACTUALIZACION PARA LOS NUMEROS Y BOOLEANOS

from datetime import date
from random import choice, randint

from gestorAplicacion.datosAleatorios import randbool, tiposVehiculos, cargosEnCocina, especialidadesChefs, randPlaca

from gestorAplicacion.usuariosRestaurante.empleado import Empleado
from gestorAplicacion.usuariosRestaurante.repartidor import Repartidor
from gestorAplicacion.usuariosRestaurante.mesero import Mesero
from gestorAplicacion.usuariosRestaurante.chef import Chef

from gestorAplicacion.gestionRestaurante.producto import Producto
from gestorAplicacion.gestionRestaurante.pedido import Pedido
from gestorAplicacion.gestionRestaurante.estadoPedido import EstadoPedido
from gestorAplicacion.gestionRestaurante.tipoPedido import TipoPedido

from excepciones.excepcionLista import ExcepcionLista

class Administrador(Empleado):

    # Atributos estaticos

    _administradores = []
    _IMPUESTOS = 1.19 # Constante

    # Constructor

    def __init__(self, cedula = 0, nombre = "", disponibilidad = False, salario = 0, restaurante = None):
        super().__init__(cedula, nombre, "Administrador", disponibilidad, salario, restaurante)

        Administrador._administradores.append(self)

    # Getters y Setters

    @classmethod
    def getAdministradores(cls):
        return cls._administradores

    @classmethod
    def setAdministradores(cls, administradores):
        cls._administradores = administradores

    @classmethod
    def getImpuestos(cls):
        return cls._IMPUESTOS

    @classmethod
    def setImpuestos(cls, impuestos):
        cls._IMPUESTOS = impuestos

    # Metodos

    # Crear un nuevo empleado (O alguno de sus subtipos) y añadirlo a los empleados del restaurante
    def contratarEmpleado(self, cedula, nombre, cargo, disponibilidad, salario, restaurante):
        empleadoNuevo = None

        # Verificar cargo del empleado
        if cargo == "Mesero":
            empleadoNuevo = Mesero(cedula, nombre, disponibilidad, salario, restaurante)
        elif cargo == "Repartidor":
            poseeVehiculo = randbool()
            placa = randPlaca()
            tipoVehiculo = choice(tiposVehiculos)
            empleadoNuevo = Repartidor(cedula, nombre, disponibilidad, salario, restaurante, poseeVehiculo, placa, tipoVehiculo)
        elif cargo == "Chef":
            cargoEnCocina = choice(cargosEnCocina)
            especialidad = choice(especialidadesChefs)
            empleadoNuevo = Chef(cedula, nombre, disponibilidad, salario, restaurante, cargoEnCocina, especialidad)
        else:
            empleadoNuevo = Empleado(cedula, nombre, disponibilidad, salario, restaurante)

        # Añadir a la lista de empleados
        listaEmpleados = self._restaurante.getEmpleados()
        listaEmpleados.append(empleadoNuevo)
        self._restaurante.setEmpleados(listaEmpleados)

        # Mensaje de confirmacion
        return f"Empleado {nombre} creado con exito"

    # Sacar empleado de la lista del restaurente, primero verificar si existe
    def despedirEmpleado(self, idEmpleado):
        listaEmpleados = self._restaurante.getEmpleados()

        # Verificar si el numero dado pertenece a la lista
        if len(listaEmpleados) > idEmpleado and idEmpleado >= 0:
            nombre = listaEmpleados[idEmpleado].getNombre()
            del listaEmpleados[idEmpleado]
            self._restaurante.setEmpleados(listaEmpleados)

            return f"El empleado {nombre} ha sido despedido"
        else:
            raise ExcepcionLista([idEmpleado, len(listaEmpleados)-1])

    # Crear nuevo producto y añadirlo al menu del restaurante
    def crearProducto(self, nombre, descripcion, precio, disponibilidad, restriccion, cantidad):
        productoNuevo = Producto(nombre, descripcion, precio, disponibilidad, restriccion, cantidad)
        
        # Verificar primero si ya existe un producto que se llame asi
        listaMenu = self._restaurante.getMenu()
        listaNombresMenu = []
        
        for producto in listaMenu:
            listaNombresMenu.append(producto.getNombre())

        # Si no existe, añadirlo a la lista
        if not nombre in listaNombresMenu:
            listaMenu.append(productoNuevo)
            self._restaurante.setMenu(listaMenu)

            return f"Producto {nombre} creado con exito"
        else:
            return f"El producto que intentas crear ya existe"

    # Actualizar el nombre de un producto, primero verificar si existe
    def actualizarNombreProducto(self, idProducto, nombre):
        # Verificar si hay ya un producto que se llame asi
        listaMenu = self._restaurante.getMenu()
        listaNombresMenu = []

        for producto in listaMenu:
            listaNombresMenu.append(producto.getNombre())
        
        # Verificar que existe el ID especificado
        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]

            # Si no hay producto con el mismo nombre, actualizarlo
            if not nombre in listaNombresMenu:
                productoActualizado.setNombre(nombre)
                listaMenu[idProducto] = productoActualizado
                self._restaurante.setMenu(listaMenu)

                return f"Producto {productoActualizado.getNombre()} actualizado con exito"
            else:
                return f"Ya exista un producto con este nombre. Intentalo de nuevo con otro nombre"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Actualizar la descripcion de un producto, primero verificar si existe
    def actualizarDescripcionProducto(self, idProducto, descripcion):
        listaMenu = self._restaurante.getMenu()

        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]
            productoActualizado.setDescripcion(descripcion)
            listaMenu[idProducto] = productoActualizado
            self._restaurante.setMenu(listaMenu)

            return f"Producto {productoActualizado.getNombre()} actualizado con exito"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Actualizar el precio de un producto, primero verificar si existe
    def actualizarPrecioProducto(self, idProducto, precio):
        listaMenu = self._restaurante.getMenu()

        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]
            productoActualizado.setPrecio(precio)
            listaMenu[idProducto] = productoActualizado
            self._restaurante.setMenu(listaMenu)

            return f"Producto {productoActualizado.getNombre()} actualizado con exito"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Actualizar la restriccion de edad de un producto, primero verificar si existe
    def actualizarRestriccionProducto(self, idProducto, restriccion):
        listaMenu = self._restaurante.getMenu()

        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]
            productoActualizado.setRestriccion(restriccion)
            listaMenu[idProducto] = productoActualizado
            self._restaurante.setMenu(listaMenu)

            return f"Producto {productoActualizado.getNombre()} actualizado con exito"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Actualizar la disponibilidad de un producto, primero verificar si existe
    def actualizarDisponibilidadProducto(self, idProducto, disponibilidad):
        listaMenu = self._restaurante.getMenu()

        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]
            productoActualizado.setDisponibilidad(disponibilidad)
            listaMenu[idProducto] = productoActualizado
            self._restaurante.setMenu(listaMenu)

            return f"Producto {productoActualizado.getNombre()} actualizado con exito"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Actualizar la cantidad disponible de un producto, primero verificar si existe
    def actualizarCantidadProducto(self, idProducto, cantidad):
        listaMenu = self._restaurante.getMenu()

        if len(listaMenu) > idProducto and idProducto >= 0:
            productoActualizado = listaMenu[idProducto]
            productoActualizado.setCantidad(cantidad)
            listaMenu[idProducto] = productoActualizado
            self._restaurante.setMenu(listaMenu)

            return f"Producto {productoActualizado.getNombre()} actualizado con exito"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Eliminar un producto del menu, primero verificar si existe
    def eliminarProducto(self, idProducto):
        listaMenu = self._restaurante.getMenu()
        
        if len(listaMenu) > idProducto and idProducto >= 0:
            producto = listaMenu[idProducto].getNombre()
            del listaMenu[idProducto]
            self._restaurante.setMenu(listaMenu)

            return f"El producto {producto} ha sido eliminado"
        else:
            raise ExcepcionLista([idProducto, len(listaMenu)-1])

    # Funcionalidad de pago de nomina
    
    def pagoNomina(self, idEmpleado = None):
        listaEmpleados = self._restaurante.getEmpleados()

        # Verificar si se especifico un empleado, si no, se le paga a todos
        if idEmpleado is None:
            total = 0

            # Conseguir el total de los salarios
            for empleado in listaEmpleados:
                total += empleado.getSalario() * Administrador.getImpuestos() # Considerar el 19% de impuestos

            # Verificar si se posee el dinero suficiente para pagar
            if total <= self._restaurante.getBalanceCuenta():
                nuevoBalance = self._restaurante.getBalanceCuenta() - total
                self._restaurante.setBalanceCuenta(nuevoBalance)

                return f"Nomina de todos los empleados pagada con exito. El nuevo balance de cuenta es: ${self._restaurante.getBalanceCuenta()}"
            else:
                return f"No se posee el suficiente dinero para pagar la nomina de todos los empleados (${total})"
        
        # Pagar a un empleado en especifico
        else:
            empleado = self._restaurante.getEmpleados()[idEmpleado]
            salario = empleado.getSalario() * Administrador.getImpuestos()

            # Verificar si existe el empleado
            if  not (len(listaEmpleados) > idEmpleado and idEmpleado >= 0):
                raise ExcepcionLista([idEmpleado, len(listaEmpleados)-1])

            # Verificar si se posee el dinero suficiente para pagar
            if salario <= self._restaurante.getBalanceCuenta():
                nuevoBalance = self._restaurante.getBalanceCuenta() - salario
                self._restaurante.setBalanceCuenta(nuevoBalance)

                return f"Nomina del empleado {empleado.getNombre()} pagada con exito. El nuevo balance de cuenta es: ${self._restaurante.getBalanceCuenta()}"
            else:
                return f"No se posee el suficiente dinero para pagar la nomina del empleado (${salario})"

    # Funcionalidad simular pedido

    def simularPedido(self, cliente):
        # Crear el nuevo pedido
        pedido = Pedido(cliente)
        pedido.setRestaurante(self._restaurante)
        pedido.setCodigo(Pedido.getTotalPedidos())
        pedido.setFechaHora(date.today().strftime("%b-%d-%Y"));

        # Escoger aleatoriamente el tipo del pedido
        tipoRandom = randint(0,2)

        if tipoRandom == 0:
            tipo = TipoPedido.LLEVAR.value
        elif tipoRandom == 1:
            tipo = TipoPedido.TIENDA.value
        elif tipoRandom == 2:
            tipo = TipoPedido.DOMICILIO.value

        pedido.setTipo(tipo)
        
        # Asignar un mesero aleatorio
        listaMeseros = Mesero.getMeseros()
        meseroElegido = listaMeseros[randint(0, len(listaMeseros)-1)]
        meseroElegido.agregarPedidoHistorial(pedido, randint(1, 10))

        # Escoger aleatoriamente una cantidad de productos del restaurante (Minimo 1)
        cantidadProductos = randint(1, len(self._restaurante.getMenu())-1)
        productos = []

        for i in range(cantidadProductos):
            # Escoger el producto de la lista
            productoRandom = randint(0, len(self._restaurante.getMenu())-1)
            producto = self._restaurante.getMenu()[productoRandom]

            # Verificar si el producto ya esta en la lista del pedido
            if producto not in productos:
                # Escoger cantidad a pedir del producto y agregarlo al pedido
                cantidadXProducto = randint(1, 5)
                producto.setCantidad(cantidadXProducto)
                productos.append(producto)
            else:
                i -= 1

        pedido.setProductos(productos)

        # Calcular el precio total
        precioTotal = 0

        for producto in productos:
            precioTotal += producto.getCantidad() * producto.getPrecio()

        pedido.setPrecioTotal(precioTotal)

        # Poner el producto en estado recibido y agregarlo a la lista del restaurante
        pedido.setEstado(EstadoPedido.RECIBIDO.value);
        self._restaurante.agregarPedido(pedido);
        return pedido

    # Implementacion de la interfaz

    def informacion(self):
        info = f"El Administrador {self._nombre} del restaurente {self._restaurante.getNombre()} es {self._nombre} con C.C. {self._cedula}.\n" \
               f"Tiene un salario de: ${self._salario}.\n"
        
        if self._disponibilidad:
            return info + "Esta disponible actualmente."
        else:
            return info + "No esta disponible actualmente."